### PR TITLE
Bug fixes for local disk rpm get on NVMe.

### PR DESCRIPTION
**Pull request by cathay4t**
~~~Don't commit yet. Will rebase after #207.~~~

 * Fix `lsmcli ldl` try-except workflow.
 * Fix local disk rpm get failure on NVMe disks.

### CODE REVIEW

File: c_binding/libsg.c [hunk 1]
```diff
@@ -297,6 +297,13 @@ int _sg_io_vpd(char *err_msg, int fd, uint8_t page_code, uint8_t *data)
                                  sense_err_msg);
             }
         }
+        /* NVMe disk does not support SCSI VPD page 0x00 which fallback to here
+         */
+        if (ioctl_errno == ENOTTY) {
```
Comment (joehandzik) [inline c_binding/libsg.c]: Is this the only error code that an NVMe device can kick back when it doesn't support 0x00?

Comment (cathay4t) [inline c_binding/libsg.c]: Yeah. Normally, if a VPD query failed, libsg will check VPD page 0x00 to confirm that page is not supported. Then we check sense data. Apparently, NVMe does not support VPD page 0x00 or SCSI sense data, hence we failback to IOCTL error code.

RHEL 7 has some patch applied to kernel to support SCSI VPD pages for NVMe, but it does not works in Fedora kernel.

**Comment by cathay4t** [comment]
@joehandzik @tasleson Please review this PR and let me know if any problem found. Thanks.

**Comment by tasleson** [comment]
@cathay4t The function `_sg_io_vpd` uses a combination of `goto out` and return statements.  For example, line 301 we are missing both (no goto, no return), so the code drops through two error message handlers, line 302 & 315.  I'm guessing we are missing a goto?

```C
269     if (ioctl_errno != 0) {
270         if (page_code == _SG_T10_SPC_VPD_SUP_VPD_PGS) {
271             _lsm_err_msg_set(err_msg, "Not a SCSI compatible device");
272             return LSM_ERR_NO_SUPPORT;
273         }
274         if (_check_sense_data(sense_err_msg, sense_data, &sense_key) != 0) {
275             if (sense_key == _T10_SPC_SENSE_KEY_ILLEGAL_REQUEST) {
276                 /* Check whether provided page is supported */
277                 rc_vpd_00 = _sg_io_vpd(err_msg, fd, _SG_T10_SPC_VPD_SUP_VPD_PGS,
278                                        vpd_00_data);
279                 if (rc_vpd_00 != 0) {
280                     rc = LSM_ERR_NO_SUPPORT;
281                     goto out;
282                 }
283                 if (_sg_is_vpd_page_supported(vpd_00_data, page_code) == true) {
284                     /* Current VPD page is supported, then it's a library bug */
285                     rc = LSM_ERR_LIB_BUG;
286                     _lsm_err_msg_set(err_msg,
287                                      "BUG: VPD page 0x%02x is supported, "
288                                      "but failed with error %d(%s), %s",
289                                      page_code, ioctl_errno,
290                                      strerror_r(ioctl_errno, strerr_buff,
291                                                 _LSM_ERR_MSG_LEN),
292                                                 sense_err_msg);
293                     goto out;
294                 } else {
295                     rc = LSM_ERR_NO_SUPPORT;
296                     _lsm_err_msg_set(err_msg, "SCSI VPD 0x%02x page is "
297                                      "not supported", page_code);
298                     goto out;
299                 }
300             } else {
301                 rc = LSM_ERR_LIB_BUG;
302                 _lsm_err_msg_set(err_msg,
303                                  "BUG: Unexpected failure of _sg_io_vpd(): %s",
304                                  sense_err_msg);

                      // Missing goto or return?

305             }
306         }
307         /* NVMe disk does not support SCSI VPD page 0x00 which fallback to here
308          */
309         if (ioctl_errno == ENOTTY) {
310             _lsm_err_msg_set(err_msg, "VPD page 0x%02x is not supported",
311                              page_code);
312             return LSM_ERR_NO_SUPPORT;
313         }
314         rc = LSM_ERR_LIB_BUG;
315         _lsm_err_msg_set(err_msg, "BUG: Unexpected failure of _sg_io_vpd(): "
316                          "error %d(%s), with no error in SCSI sense data",
317                          ioctl_errno,
318                          strerror_r(ioctl_errno, strerr_buff, _LSM_ERR_MSG_LEN)
319                          );
320     }


```


I think it would be better to adopt one approach or the other and stick to it throughout the function.

**Comment by cathay4t** [comment]
@tasleson Thanks for the review. Will fix that.

**Comment by joehandzik** [comment]
@cathay4t Yeah, aside from that bit from @tasleson, I think this is fine.

**Comment by cathay4t** [comment]
New patch uploaded.

**Comment by tasleson** [comment]
@cathay4t Thanks!

**Pull request merged**